### ISSUE TITLE
Check handling of attributes on function declaration parameters

### ISSUE THREAD

**Issue by jrfnl** [PHPCSStandards/PHP_CodeSniffer#1282]
## Setting the scene

PHP 8.0 introduced support for attributes via the `#[...]` syntax. At this moment, neither PHPCS itself, nor PHPCSExtra contain any sniffs to handle the formatting of attributes.

The [PER Coding Standard](https://www.php-fig.org/per/coding-style/) from FIG, since PER 2.0, outlines a set of rules for attribute formatting to comply with, so using those rules as a starting point would allow for creating an initial set of sniffs to address attribute formatting.

## The PER rule

[From PER](https://www.php-fig.org/per/coding-style/#122-placement):
> For attributes on parameters, if the parameter list is presented on a single line, the attribute MUST be placed inline with the parameter it describes, separated by a single space. If the parameter list is split into multiple lines for any reason, the attribute MUST be placed on its own line prior to the parameter, indented the same as the parameter. If the parameter list is split into multiple lines, a blank line MAY be included between one parameter and the attributes of the following parameter in order to aid readability.

## The action

The above rules should probably be covered via pre-existing function declaration sniffs; or at the very least, if a new sniff would be introduced to address this, it needs to be safeguarded that the pre-existing sniff(s) and the new sniff do not create fixer conflicts.

I imagine that the "may include a blank line" part of the above rule could cause fixer conflicts, but there may be more potentially conflicting aspects

It will need investigating which sniff(s) should handle this and whether these sniffs already handle attributes on parameter declarations or not.

The initial outcome should be a list of sniff names for sniffs which handle function declarations and whether these currently ignore/format parameter attributes and if so, whether how this is handled is in line with the PER rules or not.

Bonus points for analyzing the function declaration sniffs in enough detail to be able to advise whether the existing sniff(s) should be updated or whether this should be handled via a separate sniff and where the risks of fixer conflicts are.

The next step after that would be updating the sniffs where needed.

Please note: the initial investigation and the sniff updates do not have to be done by the same contributor. These are separate tasks.

**Comment by ntdiary** [comment]
I took a quick look at the issue, and I feel this would be good one for me to gradually get deeper into the project. It may take me a while to work through it, though. Leaving a comment here per `CONTRIBUTING.md` to avoid duplicate efforts. :)

**Comment by ntdiary** [comment]
After some investigation, I have tentatively grouped the related sniffs into three categories:

- Primary candidates
  These are the three function declaration sniffs that are directly related to this issue, along with the newly added sniff for attributes. Some of them already have partial support for attributes, while others likely need further updates.

- Possibly reusable function call sniffs
  These are three function call sniffs that may be reusable here. Attributes with arguments can look similar to function calls from a token perspective. And PER also mentions:
  > The attribute arguments MUST follow the same rules as defined for multiline function calls.

- Other declaration sniffs (for completeness)
  These are also related to function declarations, but they do not care much about function parameters and are unlikely to require changes.

For the first two categories, I compiled a checklist of the relevant error codes. For the third category, I only included the sniff list.

I have gone through the code of the relevant sniffs, but I still need to review some details more carefully. There is quite a lot to digest here, so please consider this a draft for now; some details may still change as I get a better understanding of the code.

For reference, my initial approach was to use `phpcs -e` to list all 204 available sniffs, then filter them with `grep -i function`, which resulted in 33 candidates. I then went through them one by one to understand their responsibilities. On the other hand, I've also been looking for sniffs that use `T_FUNCTION` and `T_ATTRIBUTE`.

**Comment by ntdiary** [comment]
### Primary candidates:
#### PEAR.Functions.FunctionDeclaration
- [ ] EmptyLine
  For multi-line declarations, a single blank line may be included between one parameter and the attributes of the following parameter.
  ``` php
  <?php
  function demo(
      $a,

      #[Foo]
      $b
  ) {}
  // 4 | ERROR | [x] Blank lines are not allowed in a multi-line function declaration
  //   |       |     (PEAR.Functions.FunctionDeclaration.EmptyLine)
  ```

- [x] Indent
  This handles the attribute's start line. As for multi-line attributes, 074622bc mentioned it should be fixed by a dedicated attribute sniff
  ```php
  <?php
  function demo(
      $a,
  #[Foo]
      $b
  ) {}
  // 4 | ERROR | [x] Multi-line function declaration not indented correctly; expected 4 spaces but found 0
  //   |       |     (PEAR.Functions.FunctionDeclaration.Indent)
  ```

<details>
<summary>Unrelated codes</summary>

- [x] SpaceAfterFunction
  `Expected 1 space after FUNCTION keyword; %s found`
- [x] SpaceBeforeOpenParen
  `Expected 0 spaces before opening parenthesis; %s found`
- [x] SpaceBeforeSemicolon
  `Expected 0 spaces before semicolon; %s found`
- [x] SpaceAfterUse
  `Expected 1 space after USE keyword; found %s`
- [x] SpaceBeforeUse
  `Expected 1 space before USE keyword; found %s`
- [x] NewlineBeforeOpenBrace
  `The closing parenthesis and the opening brace of a multi-line function declaration must be on the same line`
- [x] SpaceBeforeOpenBrace
  `There must be a single space between the closing parenthesis/return type and the opening brace of a multi-line function declaration; found %s spaces`
- [x] CloseBracketLine
  `The closing parenthesis of a multi-line %s declaration must be on a new line`
- [x] UseCloseBracketLine
  `The closing parenthesis of a multi-line use declaration must be on a new line`

</details>

#### Squiz.Functions.FunctionDeclarationArgumentSpacing

- [x] SpacingAfterOpen
  ```php
  <?php
  function demo( #[Foo]$a) {}
  // 2 | ERROR | [x] Expected 0 spaces after opening parenthesis; 1 found
  //   |       |     (Squiz.Functions.FunctionDeclarationArgumentSpacing.SpacingAfterOpen)
  ```
- [ ] SpaceBeforeComma
  There seems to be a bug here. Might cause conflicts with attribute sniff?
  ```php
  <?php
  function foo(
      int $a ,
      int $b
  ) {}
  // 3 | ERROR | [x] Expected 0 spaces between argument "$a" and comma; 1 found
  //   |       |     (Squiz.Functions.FunctionDeclarationArgumentSpacing.SpaceBeforeComma)
  // phpcbf =>
  function foo(
      int $a, int $b
  ) {}
  ```
- [ ] NoSpaceBefore
  This only checks single-line function
  ```php
  <?php
  function demo($a,#[Foo] $b) {}
  // 2 | ERROR | [x] Expected 1 space between comma and argument "$b"; 0 found
  //   |       |     (Squiz.Functions.FunctionDeclarationArgumentSpacing.NoSpaceBeforeArg)
  ```
  - [ ] should update `$typeOfNext...`? Like type hint:
    ```php
    <?php
    function demo($a,int $b) {}
    // 2 | ERROR | [x] Expected 1 space between comma and type hint "int"; 0 found
    //   |       |     (Squiz.Functions.FunctionDeclarationArgumentSpacing.NoSpaceBeforeHint)
    ``` 
- [ ] SpacingBefore
  Similar to the `NoSpaceBefore`.
  ```php
  <?php
  function demo($a,  #[Foo] $b) {}
  // 2 | ERROR | [x] Expected 1 space between comma and argument "$b"; 2 found
  //   |       |     (Squiz.Functions.FunctionDeclarationArgumentSpacing.SpacingBeforeArg)
  ```
- [ ] new SpacingAfterAttribute ✨
  `Expected 1 space between attribute and argument "%s"; %s found`
  Only for single-line functions, e.g. `function demo(#[Foo]  $a) {}`

<details>
<summary>Unrelated codes</summary>

- [x] SpacingBetween
  `Expected 0 spaces between parenthesis of function declaration; %s found`
- [x] SpacingAfterReference
  `Expected 0 spaces after reference operator for argument "%s"; %s found`
- [x] SpacingAfterVariadic
  `Expected 0 spaces after variadic operator for argument "%s"; %s found`
- [x] SpaceBeforeEquals
  `Incorrect spacing between argument "%1$s" and equals sign; expected %3$s but found %2$s`
- [x] SpaceAfterEquals
  `Incorrect spacing between default value and equals sign for argument "%s"; expected`
- [x] SpacingAfterHint
  `Expected 1 space between type hint and argument "%s"; %s found`
- [x] SpacingAfterVisibility
  `Expected 1 space after visibility modifier "%s"; %s found`
- [x] SpacingAfterSetVisibility
  `Expected 1 space after set-visibility modifier "%s"; %s found`
- [x] SpacingAfterReadonly
  `Expected 1 space after readonly modifier; %s found`
- [x] SpacingBeforeClose
  `Expected %s spaces before closing parenthesis; %s found`

</details>

#### Squiz.Functions.MultiLineFunctionDeclaration
The single-line vs multi-line logic still needs some investigation.
- [x] FirstParamSpacing
  ```php
  <?php
  function demo(#[Foo]
      $a,
      #[Foo]$b
  ) {}
  // 2 | ERROR | [x] The first parameter of a multi-line function declaration must be on the line after the
  //   |       |     opening bracket (Squiz.Functions.MultiLineFunctionDeclaration.FirstParamSpacing)
  ```
- [x] OneParamPerLine
  f39847a5 skipped over attributes
  ```php
  <?php
  function demo(
      #[Foo] $a, #[Foo] $b
  ) {}
  // 3 | ERROR | [x] Multi-line function declarations must define one parameter per line
  //   |       |     (Squiz.Functions.MultiLineFunctionDeclaration.OneParamPerLine)
  ```

<details>
<summary>Unrelated codes</summary>

- [x] CloseBracketNewLine
  `There must not be a newline before the closing parenthesis of a single-line function declaration`

</details>

#### a dedicated sniff for attributes ✨
This sniff could handle these cases:
```php
<?php
// name must immediately follow the opening attribute block indicator
function demo(#[ Foo] $a) {}

// If an attribute has no arguments, the () must be omitted.
function demo(#[Foo()] $a) {}

// Expected 0 spaces between last character of attribute name and "]"
function demo(#[Foo ] $a) {}

// Expected 0 spaces between ")" and "]"
function demo(#[Foo(1) ] $a) {}

// Expected 0 spaces before comma
function demo(#[Foo , Bar] $a) {}

// Expected 1 spaces after comma
function demo(#[Foo,Bar] $a) {}

// Multi-line Attributes, Bar must be placed in separate attribute blocks
function demo(
    #[
        Foo,
        Bar
    ]
    $a
) {}

// Attributes must be placed on its own line prior to the parameter
function demo(
    #[Foo] $a,
    $b
)

// Attribute blocks must be on separate lines with no blank lines between them
function demo(
    #[Foo] #[Bar]
    $a,
    $b
)
```

**Comment by ntdiary** [comment]
### Possibly reusable function call sniffs
#### Generic.Functions.FunctionCallArgumentSpacing
- [ ] SpaceBeforeComma
  ```php
  <?php
  function demo(#[Foo(1 , 2)] $a) {}
  // 2 | ERROR | [x] Space found before comma in argument list
  //   |       |     (Generic.Functions.FunctionCallArgumentSpacing.SpaceBeforeComma)
  ```
- [ ] NoSpaceAfterComma
  ```php
  <?php
  function demo(#[Foo(1,2)] $a) {}
  // 2 | ERROR | [x] No space found after comma in argument list
  //   |       |     (Generic.Functions.FunctionCallArgumentSpacing.NoSpaceAfterComma)
  ```
- [ ] TooMuchSpaceAfterComma
  ```php
  <?php
  function demo(#[Foo(1,  2)] $a) {}
  // 2 | ERROR | [x] Expected 1 space after comma in argument list; 2 found
  //   |       |     (Generic.Functions.FunctionCallArgumentSpacing.TooMuchSpaceAfterComma)
  ```

#### PEAR.Functions.FunctionCallSignature
- [ ] SpaceBeforeOpenBracket
  ```php
  <?php
  function demo(#[Foo (1)] $a) {}
  // 2 | ERROR | [x] Space before opening parenthesis of function call prohibited
  //   |       |     (PEAR.Functions.FunctionCallSignature.SpaceBeforeOpenBracket)
  ```
- [ ] SpaceAfterCloseBracket
  `Space after closing parenthesis of function call prohibited`
  For now, only works if the next token is `T_SEMICOLON`

- [ ] SpaceAfterOpenBracket
  has a property `requiredSpacesAfterOpen`
  ```php
  <?php
  function demo1(#[Foo( 1)] $a) {}
  // phpcs:set PEAR.Functions.FunctionCallSignature requiredSpacesAfterOpen 1
  function demo2(#[Foo(  1)] $a) {}
  // 2 | ERROR | [x] Space after opening parenthesis of function call prohibited
  //   |       |     (PEAR.Functions.FunctionCallSignature.SpaceAfterOpenBracket)
  // 4 | ERROR | [x] Expected 1 spaces after opening parenthesis; 2 found
  //   |       |     (PEAR.Functions.FunctionCallSignature.SpaceAfterOpenBracket)
  ```
- [ ] SpaceBeforeCloseBracket
  has a property `requiredSpacesBeforeClose`
  ```php
  <?php
  function demo(#[Foo(1 )] $a) {}
  // 2 | ERROR | [x] Expected 0 spaces before closing parenthesis; 1 found
  //   |       |     (PEAR.Functions.FunctionCallSignature.SpaceBeforeCloseBracket)
  ```
- [ ] OpeningIndent
  ```php
  <?php
  function demo(
    #[Foo(
        1, 2
    )]
    $a
  ) {}
  // 3 | ERROR | [x] Opening statement of multi-line function call not indented correctly; expected 0 spaces
  //   |       |     but found 2 (PEAR.Functions.FunctionCallSignature.OpeningIndent)
  ```
- [ ] ContentAfterOpenBracket
  ```php
  <?php
  function demo(
  #[Foo(1,
      2
  )]
    $a
  ) {}
  // 3 | ERROR | [x] Opening parenthesis of a multi-line function call must be the last content on the line
  //   |       |     (PEAR.Functions.FunctionCallSignature.ContentAfterOpenBracket)
  ```
- [ ] CloseBracketLine
  ```php
  <?php
  function demo(
  #[Foo(
      1,
  2)]
    $a
  ) {}
  // 5 | ERROR | [x] Closing parenthesis of a multi-line function call must be on a line by itself
  //   |       |     (PEAR.Functions.FunctionCallSignature.CloseBracketLine)
  ```
- [ ] FirstArgumentPosition
  ```php
  <?php
  function demo(
  #[Foo(
  
      1,
      2
  )]
    $a
  ) {}
  // 5 | ERROR | [x] The first argument in a multi-line function call must be on the line after the opening
  //   |       |     parenthesis (PEAR.Functions.FunctionCallSignature.FirstArgumentPosition)
  ```
- [ ] EmptyLine
  ```php
  <?php
  function demo(
  #[Foo(
      1,
  
      2
  )]
    $a
  ) {}
  // 5 | ERROR | [x] Empty lines are not allowed in multi-line function calls
  //   |       |     (PEAR.Functions.FunctionCallSignature.EmptyLine)
  ```
- [ ] Indent
  ```php
  <?php
  function demo(
  #[Foo(
      1,
        2
  )]
    $a
  ) {}
  // 5 | ERROR | [x] Multi-line function call not indented correctly; expected 4 spaces but found 6
  //   |       |     (PEAR.Functions.FunctionCallSignature.Indent)
  ```
- [ ] MultipleArguments
  ```php
  <?php
  // phpcs:set PEAR.Functions.FunctionCallSignature allowMultipleArguments false
  function demo(
  #[Foo(
      1,2
  )]
    $a
  ) {}
  // 5 | ERROR | [x] Only one argument is allowed per line in a multi-line function call
  //   |       |     (PEAR.Functions.FunctionCallSignature.MultipleArguments)
  ```

#### PSR2.Methods.FunctionCallSignature
`extends PEARFunctionCallSignatureSniff`
Changes `allowMultipleArguments` to `false` and the `isMultiLineCall` logic

**Comment by ntdiary** [comment]
### Other declaration sniffs (for completeness)
- Generic.Functions.OpeningFunctionBraceBsdAllman
  Checks that the opening brace of a function is on the line after the function declaration.
- Generic.Functions.OpeningFunctionBraceKernighanRitchie
  Checks that the opening brace of a function is on the same line as the function declaration.
- PSR12.Functions.NullableTypeDeclaration
  Verifies that nullable typehints are lacking superfluous whitespace, e.g. ?int
- Squiz.Functions.FunctionDeclaration
  Checks the function declaration is correct (ignoring parameter parts).
- Squiz.WhiteSpace.FunctionClosingBraceSpace
  Checks that there is one empty line before the closing brace of a function.
- Squiz.WhiteSpace.FunctionSpacing
  Checks the separation between functions and methods.
- Squiz.WhiteSpace.FunctionOpeningBraceSpace
  Checks that there is no empty line after the opening brace of a function.
